// easybase-react/dist/SignIn-2d872e4e.js
var React = require('react');
var index_esm = require('./index.esm-264e892f.js');
var NativeAuth = require('./NativeAuth-528322b5.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _require = require('easybase-react'),
    useEasybase = _require.useEasybase;

function SignIn (_ref) {
  var setCurrentPage = _ref.setCurrentPage,
      dictionary = _ref.dictionary,
      toast = _ref.toast;

  var _useForm = index_esm.useForm(),
      control = _useForm.control,
      handleSubmit = _useForm.handleSubmit,
      reset = _useForm.reset,
      isSubmitting = _useForm.formState.isSubmitting;

  var _useEasybase = useEasybase(),
      signIn = _useEasybase.signIn;

  var onSubmit = function onSubmit(formData) {
    try {
      if (!formData.email || !formData.password) {
        return Promise.resolve();
      }

      return Promise.resolve(signIn(formData.email, formData.password)).then(function (signInRes) {
        if (!signInRes.success) {
          if (signInRes.errorCode === "NoUserExists") {
            toast(dictionary.errorUserDoesNotExist);
          } else if (signInRes.errorCode === "BadFormat") {
            reset();
            toast(dictionary.errorBadInputFormat);
          }
        }
      }); // Will automatically change views
    } catch (e) {
      return Promise.reject(e);
    }
  };

  return React__default["default"].createElement(NativeAuth.Form, null, React__default["default"].createElement(NativeAuth.HeaderText, null, dictionary.signInHeader), React__default["default"].createElement(NativeAuth.MainView, null, React__default["default"].createElement(index_esm.Controller, {
    control: control,
    render: function render(_ref2) {
      var _ref2$field = _ref2.field,
          onChange = _ref2$field.onChange,
          onBlur = _ref2$field.onBlur,
          value = _ref2$field.value;
      return React__default["default"].createElement(NativeAuth.Input, {
        onBlur: onBlur,
        onChangeText: function onChangeText(value) {
          return onChange(value);
        },
        value: value,
        placeholder: dictionary.emailLabel,
        editable: !isSubmitting,
        keyboardType: "email-address",
        returnKeyType: "default",
        autoCompleteType: "email",
        autoCapitalize: "none"
      });
    },
    name: "email",
    defaultValue: ""
  }), React__default["default"].createElement(NativeAuth.SpacerXL, null), React__default["default"].createElement(index_esm.Controller, {
    control: control,
    render: function render(_ref3) {
      var _ref3$field = _ref3.field,
          onChange = _ref3$field.onChange,
          onBlur = _ref3$field.onBlur,
          value = _ref3$field.value;
      return React__default["default"].createElement(NativeAuth.Input, {
        onBlur: onBlur,
        onChangeText: function onChangeText(value) {
          return onChange(value);
        },
        value: value,
        placeholder: dictionary.passwordLabel,
        editable: !isSubmitting,
        returnKeyType: "default",
        secureTextEntry: true,
        autoCompleteType: "password",
        autoCapitalize: "none"
      });
    },
    name: "password",
    defaultValue: ""
  }), React__default["default"].createElement(NativeAuth.ForgotPassword, {
    onPress: function onPress(_) {
      return setCurrentPage("ForgotPassword");
    },
    disabled: isSubmitting,
    title: dictionary.forgotPasswordButton
  })), React__default["default"].createElement(NativeAuth.View, null, React__default["default"].createElement(NativeAuth.SubmitButton, {
    onPress: handleSubmit(onSubmit),
    disabled: isSubmitting,
    title: dictionary.signInSubmitButton
  }), React__default["default"].createElement(NativeAuth.SpacerS, null), React__default["default"].createElement(NativeAuth.SecondaryButton, {
    onPress: function onPress(_) {
      return setCurrentPage("SignUp");
    },
    disabled: isSubmitting,
    title: dictionary.noAccountButton
  })));
}

exports["default"] = SignIn;
