// easybase-react/dist/SignIn-e3dbf321.js
import React__default from 'react';
import { T as TextButton, u as useForm, F as Form, H as HeaderText, S as Spacer, E as EmailInput, P as PasswordInput, a as SubmitButton, b as SecondaryButton } from './index.esm-189a9937.js';
import { _ as _extends, u as useEasybase } from './index-775190d4.js';
import { s as styled, t as toast } from './Auth-25393280.js';

const ForgotPassword = styled(TextButton)(props => _extends({
  marginTop: -53,
  marginBottom: 53,
  display: 'flex'
}, props.theme.forgotPassword ? _extends({}, props.theme.forgotPassword) : {}));
function ForgotPassword$1 (props) {
  return React__default.createElement(ForgotPassword, Object.assign({}, props, {
    type: "button"
  }));
}

function SignIn ({
  setCurrentPage,
  dictionary
}) {
  const {
    register,
    handleSubmit,
    reset,
    formState: {
      isSubmitting
    }
  } = useForm();
  const {
    signIn
  } = useEasybase();

  const onSubmit = async formData => {
    await new Promise(resolve => setTimeout(resolve, 1000));
    const signInRes = await signIn(formData.email, formData.password);

    if (!signInRes.success) {
      if (signInRes.errorCode === "NoUserExists") {
        toast.error(dictionary.errorUserDoesNotExist);
      } else if (signInRes.errorCode === "BadFormat") {
        reset();
        toast.error(dictionary.errorBadInputFormat);
      }
    } // Will automatically change views

  };

  return React__default.createElement(Form, {
    onSubmit: handleSubmit(onSubmit)
  }, React__default.createElement(HeaderText, null, dictionary.signInHeader), React__default.createElement(Spacer, {
    size: "medium"
  }), React__default.createElement(EmailInput, {
    register: () => register("email"),
    label: dictionary.emailLabel,
    disabled: isSubmitting
  }), React__default.createElement(Spacer, {
    size: "xlarge"
  }), React__default.createElement(PasswordInput, {
    register: () => register("password"),
    autoComplete: "current-password",
    disabled: isSubmitting,
    label: dictionary.passwordLabel
  }), React__default.createElement(Spacer, {
    size: "xlarge"
  }), React__default.createElement(ForgotPassword$1, {
    onClick: _ => setCurrentPage("ForgotPassword"),
    disabled: isSubmitting
  }, dictionary.forgotPasswordButton), React__default.createElement(SubmitButton, {
    disabled: isSubmitting
  }, dictionary.signInSubmitButton), React__default.createElement(SecondaryButton, {
    onClick: _ => setCurrentPage("SignUp"),
    disabled: isSubmitting
  }, dictionary.noAccountButton));
}

export { SignIn as default };
